doc/Route_priority_and_TTL.md

```diff
@@ -0,0 +1,137 @@
+
+In this design, the message routing and queue layers remain agnostic of the underlying endpoint implementations. This means even though message priorities primarily target the IoT Hub endpoint, message priorities on brokered endpoints is also enabled.
+
+###	Handling priorities
+The endpoint executor takes the provided context and expand the collection of sequential stores to be one per priority per endpoint.  When a new message comes in, the executor enqueues it into the sequential store that matches the endpoint and priority.  For naming, these new sequential stores take the endpoint ID, and append “_Pri<value>” to the end of the string, where <value> is the message priority this sequential store represents (e.g. “iothub_Pri0”, “iothub_Pri1”, etc).
+
+![](./images/route_priority_ttl/priority_message_pump_components.jpg)
+
+On the message processing side, the message pump iterates messages from each sequential store in priority order.  At the end of every loop, the pump always reset back to priority 0 (or whichever is the highest) again.  This is so that high priority messages that come in while low priority messages are in-flight can still be processed in a reasonable timeframe.  This means low priority message can potentially be starved if the throughput is not enough to service all the higher priority messages.  However, this behavior is desirable based on known customer scenarios.
+
+![](./images/route_priority_ttl/priority_message_pump_flowchart.jpg)
+
+The message pump does not attempt to aggregate messages of different priorities into a single batch.  This is to guarantee that only one type of priority is processed at a time, which greatly simplifies the checkpointing requirements.
+To process the messages from a given priority, a corresponding instance of Checkpointer is needed to keep track of failures and retries.  To accomplish this, the executor maintains a map of priorities to FSMs, the latter of which now contains the checkpointer specific to its queue.
+
+### Handling TTL
+As mentioned above, the TTL specific to a route is now available when a message gets stored.  This TTL value is a new property on the stored MessageRef object, so that it can be referenced when the message pump processes the message.
+
+As the message pump fetches messages out of the store, the MessageStore checks the message enqueue timestamp against the TTL timespan on the message for expiration.  If the message has no TTL, then global TTL will be used.
+
+If the message is expired, it simply gets skipped over, and GetNext() keeps on iterating until it fills the requested batch size or the store is empty.  Then in the background, the existing message cleanup task will periodically delete these expired messages.
```

Comment:
we have a flag that controls aggressive vs normal cleanup. Maybe we should document that as well.

Reply:
Dylan wrote up a section on this and I added it in.